consensus/service/src/timestamp_validator.rs

```diff
@@ -0,0 +1,236 @@
+// Copyright (c) 2023 The MobileCoin Foundation
+
+use displaydoc::Display;
+use mc_blockchain_types::Block;
+
+/// Provides logic for validating a timestamp used in consensus
+
+const MAX_TIMESTAMP_AGE: u64 = 30 * 1000; // 30 seconds
+
+/// Errors related to validating timestamps
+#[derive(Clone, Debug, Display, Eq, PartialEq)]
+pub enum Error {
+    /// The timestamp is in the future now: {0}, timestamp: {1}
+    InFuture(u64, u64),
+    /** The timestamp is not newer than the last bock. last_bock timestamp: {0},
+    timestamp: {1} */
+    NotNewerThanLastBlock(u64, u64),
+    /// The timestamp is too far in the past now: {0}, timestamp: {1}
+    TooOld(u64, u64),
+}
+
+pub fn validate(timestamp: u64, latest_block: &Block) -> Result<(), Error> {
+    let now = std::time::SystemTime::now()
+        .duration_since(std::time::UNIX_EPOCH)
+        .expect("Failed to get system time")
+        .as_millis() as u64;
+
+    if timestamp > now {
```

Comment:
Up for suggestions here, we could allow for timestamps in the future.
i.e. node A's clock is a bit behind node B's so when Node A validates a value from NodeB it could seem like it's in the future. We might consider allowing 1-3 seconds ahead.
Based on looking at the historical signed at timestamps, if we throw out significant outliers all the node times are within 3 seconds of each other, https://github.com/nick-mobilecoin/timestamp-metrics/blob/main/timestamps.ipynb
Search for "We can see now that the nodes are all within 3 seconds of each other.", Using a 3 second window would throw out 80,000 blocks of the 2 million that were analyzed.